painters/model_def_standard.py

```diff
@@ -0,0 +1,137 @@
+        return categorical_crossentropy
+
+    def batch_size(self):
+        return self.my_batch_size
+
+    # This is an abstract function which the trial class needs for instantiation.
+    def validation_metrics(self):
+        return {}
+
+    # HACK: compute the validation metrics in a customized way.
+    def compute_validation_metrics(self, step_id):
+        assert self.validation_data_adapter is not None
+        assert self.model is not None
+
+        self.validation_data_adapter.start(is_validation=True)
+        validation_iterator = self.validation_data_adapter.get_iterator()
+        assert validation_iterator is not None
+
+        X_val_embedded = None
+        y_val = None
+        num_inputs = 0
+
+        # TODO: pre-allocate array for X_val_embedded and y_val
+        #  to make the computation faster
+        # Shape of X_batch: (96=batch_size, 3, 256, 256 (dims of a single image)).
+        # Shape of y_batch: (96=batch_size, number of classes).
+        for X_batch, y_batch in validation_iterator:
+            # Shape of X_embedded: (96=batch_size, number of classes).
+            X_embedded = self.model.predict(X_batch)
+
+            # Shape of X_val_embedded: (iteration * batch_size, number of classes).
+            if X_val_embedded is None:
+                X_val_embedded = X_embedded
+            else:
+                X_val_embedded = np.concatenate((X_val_embedded, X_embedded), axis=0)
+
+            # Shape of y_val: number of (iteration * batch_size, number of classes).
+            if y_val is None:
+                y_val = y_batch
+            else:
+                y_val = np.concatenate((y_val, y_batch), axis=0)
+            num_inputs += len(X_batch)
+
+        # Calculate categorical cross entropy for validation data,
+        # which is the same loss as used in training
+        y_pred_vec = tf.convert_to_tensor(X_val_embedded)
+        y_val_vec = tf.convert_to_tensor(y_val)
+        cce = categorical_crossentropy(y_pred_vec, y_val_vec)
+        sess = tf.Session()
+        with sess.as_default():
+            cce = cce.eval()
+        cce = np.mean(cce)
+
+        # Calculate accuracy of single image artist classification.
+        # Class prediction is the class with max prob in vector
+        # prediction of each painting.
+        y_val = np.argmax(y_val, axis=1)
+        y_pred = np.argmax(X_val_embedded, axis=1)
+        single_painting_acc = accuracy_score(y_val, y_pred)
+
+        # Pairwise evaluation: whether they are from the same artist
+        batches_val = _create_pairs_generator(
+            X_val_embedded, y_val, lambda u, v: [u, v],
+            num_groups=32,
+            batch_size=10000)
+
+        y_pred, y_true = np.array([]), np.array([])
+        for X, y in batches_val:
+            y_pred = np.hstack((y_pred, pairs_dot(X)))
+            y_true = np.hstack((y_true, y))
+        roc_auc = roc_auc_score(y_true, y_pred)
+
+        self.validation_data_adapter.stop()
```

Comment:
You can stop this earlier (around line 86), when you're done using the `validation_iterator` object.